src/Model/Behavior/SearchBehavior.php

```diff
@@ -167,19 +159,23 @@ protected function _extractParams($params, $filters)
      * ```
      *
      * @param array $params The parameters array to flatten.
-     * @param array $flattenWhitelist Keys to avoid flattening.
+     * @param array $filters The array of filters with configuration
      * @return array The flattened parameters array.
      */
-    protected function _flattenParams($params, $flattenWhitelist)
-    {
+    protected function _flattenParams($params, $filters)
+    {  
         $flattened = [];
         foreach ($params as $key => $value) {
-            if (is_array($value) && !in_array($key, $flattenWhitelist)) {
-                foreach ($value as $childKey => $childValue) {
-                    if (!is_numeric($childKey)) {
-                        $flattened[$key . '.' . $childKey] = $childValue;
-                    } else {
-                        $flattened[$key][$childKey] = $childValue;
+            if (is_array($value)) {
+                if (!empty($filters[$key]) && $filters[$key]->getConfig()['flatten'] == false) {
+                    $flattened[$key] = $value;
+                } else {
```

Comment:
The second `is_array()` check is necessary to ensure the original behavior is maintained, where an array passed with no filter configuration is flattened as a matter of course. Here is how the original method looked before I changed anything:
```php
protected function _flattenParams($params)
{
    $flattened = [];
    foreach ($params as $key => $value) {
        if (is_array($value)) {
            foreach ($value as $childKey => $childValue) {
                if (!is_numeric($childKey)) {
                    $flattened[$key . '.' . $childKey] = $childValue;
                } else {
                    $flattened[$key][$childKey] = $childValue;
                }
            }
        } else {
            $flattened[$key] = $value;
        }
    }
    return $flattened;
}
```
 I think the complexity has only increased minimally, and is no more nested than before.